app/javascript/pages/UtmLinks/Index.tsx

```diff
@@ -87,71 +97,95 @@ const UtmLinkList = () => {
+    }
+    if (params.query !== undefined && params.query.length > 0) {
+      newParams.query = params.query;
+    }
+    if (params.sort !== undefined && params.sort !== null) {
+      newParams.key = params.sort.key;
+      newParams.direction = params.sort.direction;
+    }
+
+    setIsLoading(true);
+    router.get(
+      Routes.utm_links_dashboard_path(newParams),
+      {},
+      {
+        preserveState: true,
+        preserveScroll: true,
+        onFinish: () => setIsLoading(false),
+      },
+    );
+  };
 
   const onChangePage = (newPage: number) => {
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      params.set("page", newPage.toString());
-      return params;
-    });
+    navigateWithParams({ page: newPage, query, sort });
   };
 
   const onSetSort = (newSort: Sort<SortKey> | null) => {
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      if (pagination.pages >= 1) params.delete("page");
-      if (newSort) {
-        params.set("key", newSort.key);
-        params.set("direction", newSort.direction);
-      }
-      return params;
-    });
     setSort(newSort);
+    navigateWithParams({ page: 1, query, sort: newSort });
   };
 
   const thProps = useSortingTableDriver<SortKey>(sort, onSetSort);
 
-  const query = searchParams.get("query") ?? "";
+  const debouncedSearch = useDebouncedCallback((newQuery: string) => {
+    navigateWithParams({ page: 1, query: newQuery, sort });
+  }, 500);
 
-  const onSearch = useDebouncedCallback((newQuery: string) => {
-    if (query === newQuery) return;
+  const onSearch = (newQuery: string) => {
+    setQuery(newQuery);
+    debouncedSearch(newQuery);
+  };
 
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      if (newQuery.length > 0) {
-        params.set("query", newQuery);
-      } else {
-        params.delete("query");
-      }
-      params.delete("page");
-      return params;
+  const handleDelete = () => {
+    if (!deletingUtmLink) return;
+
+    setDeletingUtmLink({ ...deletingUtmLink, state: "deleting" });
+    router.delete(`${Routes.utm_links_dashboard_path()}/${deletingUtmLink.id}`, {
+      only: ["utm_links", "pagination", "flash"],
+      onSuccess: () => {
+        setDeletingUtmLink(null);
+        setSelectedUtmLink(null);
+      },
+      onError: () => {
+        setDeletingUtmLink((previous) => (previous ? { ...previous, state: "delete-confirmation" } : previous));
+        showAlert("Failed to delete link. Please try again.", "error");
+      },
     });
-  }, 500);
+  };
 
   return (
     <AnalyticsLayout
       selectedTab="utm_links"
       actions={
         <>
           <SearchBoxPopover initialQuery={query} onSearch={onSearch} />
-          <Link to="/dashboard/utm_links/new" className="button accent">
+          <Link href={`${Routes.utm_links_dashboard_path()}/new`} className="button accent">
```

Comment:
Use the dedicated url helper
```suggestion
          <Link href={Routes.new_utm_link_dashboard_path()} className="button accent">
```